File: analytical_engine/java/graphx-on-graphscope/src/main/java/com/alibaba/graphscope/factory/GraphXFactory.java
package com.alibaba.graphscope.factory;

import com.alibaba.graphscope.conf.GraphXConf;
import com.alibaba.graphscope.graph.EdgeContextImpl;
import com.alibaba.graphscope.graph.GraphXVertexIdManager;
import com.alibaba.graphscope.graph.GraphxEdgeManager;
import com.alibaba.graphscope.graph.VertexDataManager;
import com.alibaba.graphscope.graph.VertexIdManager;
import com.alibaba.graphscope.graph.impl.GraphXVertexIdManagerImpl;
import com.alibaba.graphscope.graph.impl.GraphxEdgeManagerImpl;
import com.alibaba.graphscope.graph.impl.VertexDataManagerImpl;
import com.alibaba.graphscope.graphx.GSEdgeTriplet;
import com.alibaba.graphscope.graphx.SerializationUtils;
import com.alibaba.graphscope.mm.MessageStore;
import com.alibaba.graphscope.mm.impl.DefaultMessageStore;
import com.alibaba.graphscope.mm.impl.ParallelMessageStore;
import com.alibaba.graphscope.utils.GraphXProxy;
import org.apache.spark.graphx.EdgeTriplet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.Function1;
import scala.Function2;
import scala.Function3;
import scala.Tuple2;
import scala.collection.Iterator;

/**
 * Providing factory methods for creating adaptor and adptor context.
 */
public class GraphXFactory {

    private static Logger logger = LoggerFactory.getLogger(GraphXFactory.class.getName());

    public static <VD, ED, MSG> GraphXConf<VD, ED, MSG> createGraphXConf(
        Class<? extends VD> vdClass, Class<? extends ED> edClass, Class<? extends MSG> msgClass) {
        return new GraphXConf<VD, ED, MSG>(vdClass, edClass, msgClass);
    }

    private static <VD, ED, MSG> GraphXProxy<VD, ED, MSG> createGraphXProxy(
        GraphXConf<VD, ED, MSG> conf, Function3<Long, VD, MSG, VD> vprog,
        Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG>>> sendMsg,
        Function2<MSG, MSG, MSG> mergeMsg, int numCores) {
        return new GraphXProxy<VD, ED, MSG>(conf, vprog, sendMsg, mergeMsg, numCores);
    }

    public static <VD, ED, MSG> GraphXProxy<VD, ED, MSG> createGraphXProxy(
        GraphXConf<VD, ED, MSG> conf, String vprogFilePath, String sendMsgFilePath,
        String mergeMsgFilePath, int numCores) {
        Function3<Long, VD, MSG, VD> vprog = deserializeVprog(vprogFilePath, conf);
        Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG>>> sendMsg = deserializeSendMsg(
            sendMsgFilePath, conf);
        Function2<MSG, MSG, MSG> mergeMsg = deserializeMergeMsg(mergeMsgFilePath, conf);
        logger.info("deserialization success: {}, {}, {}", vprog, sendMsg, mergeMsg);

        GraphXProxy<VD, ED, MSG> graphXProxy = createGraphXProxy(conf, vprog, sendMsg, mergeMsg, numCores);
        logger.info("Construct graphx proxy: " + graphXProxy);
        return graphXProxy;
    }

    public static GraphXVertexIdManager createIdManager(GraphXConf conf) {
        return new GraphXVertexIdManagerImpl(conf);
    }

    public static <VD,ED,MSG> VertexDataManager createVertexDataManager(GraphXConf<VD,ED,MSG> conf) {
        return new VertexDataManagerImpl<VD>(conf);
    }

    public static <VD,ED,MSG> MessageStore<MSG,VD> createDefaultMessageStore(GraphXConf<VD,ED,MSG> conf) {
        return new DefaultMessageStore<MSG,VD>(conf);
    }
    public static <VD,ED,MSG> MessageStore<MSG,VD> createParallelMessageStore(GraphXConf<VD,ED,MSG> conf) {
        return new ParallelMessageStore<MSG,VD>(conf);
    }

    public static <VD, ED, MSG_T> EdgeContextImpl<VD, ED, MSG_T> createEdgeContext(
        GraphXConf<VD,ED,MSG_T> conf) {
        return new EdgeContextImpl<>(conf);
    }

    public static <VD, ED, MSG_T> GraphxEdgeManager<VD, ED, MSG_T> createEdgeManager(
        GraphXConf<VD,ED,MSG_T> conf, VertexIdManager<Long, Long> idManager,
        VertexDataManager<VD> vertexDataManager) {
        return new GraphxEdgeManagerImpl<>(conf, idManager, vertexDataManager);
    }

    public static <VD,ED>GSEdgeTriplet<VD,ED> createEdgeTriplet(GraphXConf<VD,ED,?> conf){
        return new GSEdgeTriplet<>();
    }

    private static <VD, ED, MSG> Function3<Long, VD, MSG, VD> deserializeVprog(String vprogFilePath,
        GraphXConf<VD, ED, MSG> conf) {
        try {
            Function3<Long, VD, MSG, VD> res = (Function3<Long, VD, MSG, VD>) SerializationUtils.read(
                vprogFilePath);
            return res;
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("deserialization vprog failed");
        }
    }

    private static <VD, ED, MSG> Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG>>> deserializeSendMsg(
        String sendMsgFilePath, GraphXConf<VD, ED, MSG> conf) {
        try {
            Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG>>> res = (Function1<EdgeTriplet<VD, ED>, Iterator<Tuple2<Long, MSG>>>) SerializationUtils.read(
                sendMsgFilePath);
            return res;
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("deserialization send msg failed");
        }
    }

    private static <VD, ED, MSG> Function2<MSG, MSG, MSG> deserializeMergeMsg(
        String mergeMsgFilePath, GraphXConf<VD, ED, MSG> conf) {
        try {
            Function2<MSG, MSG, MSG> res = (Function2<MSG, MSG, MSG>) SerializationUtils.read(
                mergeMsgFilePath);
            return res;
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("deserialization merge msg failed");
        }
    }
}

File: analytical_engine/java/graphx-on-graphscope/src/main/scala/org/apache/spark/graphx/GraphXFactory.scala
package org.apache.spark.graphx

import com.alibaba.graphscope.fragment.IFragment
import com.alibaba.graphscope.graphx.{GSEdgeTriplet, GSEdgeTripletImpl, SerializationUtils}
import com.alibaba.graphscope.parallel.DefaultMessageManager
import org.apache.spark.graphx.impl.graph.GraphXProxy
import org.apache.spark.graphx.impl.graph.{EdgeManagerImpl, GraphXProxy, GraphXVertexIdManagerImpl, VertexDataManagerImpl}
import org.apache.spark.graphx.impl.message.DefaultMessageStore
import org.apache.spark.graphx.traits.{GraphXVertexIdManager, VertexDataManager}
import org.apache.spark.internal.Logging

import scala.reflect.ClassTag

object GraphXFactory extends Logging{

  def createGraphxConf[VD :ClassTag,ED : ClassTag]: GraphXConf[VD, ED] = {
    new GraphXConf[VD,ED]
  }

  def createGraphxProxy[VD : ClassTag,ED : ClassTag,MSG: ClassTag](conf : GraphXConf[VD,ED], iFragment: IFragment[Long,Long,_,_], mm : DefaultMessageManager, vprogPath : String,
                                   sendMsgFilePath : String, mergeMsgFilePath : String, vdataFilePath : String, maxIter: Int,
                                   numCores : Int, vdataSize : Long, initialmessage : MSG): GraphXProxy[VD,ED,MSG] ={

    val vprog: (VertexId,VD,MSG)=>VD = deserializeVprog(vprogPath, conf)
    val sendMsg: (EdgeTriplet[VD,ED]=>Iterator[(VertexId,MSG)]) = deserializeSendMsg(sendMsgFilePath, conf)
    val mergeMsg: (MSG,MSG)=>MSG = deserializeMergeMsg(mergeMsgFilePath, conf)
    log.info(s"deserialization success: ${vprog}, ${sendMsg}, ${mergeMsg}")

    new GraphXProxy[VD,ED,MSG](conf, iFragment, mm, initialmessage,vprog, sendMsg, mergeMsg, maxIter, numCores, vdataFilePath, vdataSize)
  }

  def createVertexIdManager[VD,ED](conf: GraphXConf[VD,ED], fragment : IFragment[Long,Long,_,_]): GraphXVertexIdManager ={
    new GraphXVertexIdManagerImpl(conf, fragment)
  }

  def createVertexDataManager[VD: ClassTag, ED : ClassTag](conf: GraphXConf[VD, ED],  fragment : IFragment[Long,Long,_,_], values : Array[VD] = null,defaultOutData : VD = null.asInstanceOf[VD]): VertexDataManager[VD] = {
    new VertexDataManagerImpl[VD,ED](conf, values, fragment, defaultOutData)
  }

  def createDefaultMessageStore[VD : ClassTag,MSG : ClassTag](conf: GraphXConf[_,_], fragment : IFragment[Long,Long,_,_], idManager : GraphXVertexIdManager, vertexDataManager : VertexDataManager[VD], mergeMsg : (MSG,MSG) => MSG): DefaultMessageStore[MSG] = {
    new DefaultMessageStore[MSG]
  }
  def createEdgeTriplet[VD: ClassTag, ED : ClassTag](conf: GraphXConf[VD,ED]): GSEdgeTriplet[VD, ED] = {
    new GSEdgeTripletImpl[VD,ED]
  }

  def createEdgeManager[VD : ClassTag,ED : ClassTag](conf: GraphXConf[VD,ED], fragment : IFragment[Long,Long,_,_], vertexIdManager: GraphXVertexIdManager, vertexDataManager: VertexDataManager[VD], numCores : Int): EdgeManagerImpl[VD, ED] = {
    new EdgeManagerImpl[VD,ED](conf, fragment, vertexIdManager, vertexDataManager, numCores)
  }


  private def deserializeVprog[VD, ED, MSG](vprogFilePath: String, conf: GraphXConf[VD, ED]) : (Long,VD,MSG) => VD = {
    try {
      val res = SerializationUtils.read(vprogFilePath).asInstanceOf[(Long, VD, MSG) => VD]
      res
    } catch {
      case e: ClassNotFoundException =>
        e.printStackTrace()
        throw new IllegalStateException("deserialization vprog failed")
    }
  }

  private def deserializeSendMsg[VD, ED, MSG](sendMsgFilePath: String, conf: GraphXConf[VD, ED]) :(EdgeTriplet[VD, ED]) => Iterator[(VertexId, MSG)] = {
    try {
      val res = SerializationUtils.read(sendMsgFilePath).asInstanceOf[EdgeTriplet[VD, ED] => Iterator[(VertexId, MSG)]]
      res
    } catch {
      case e: ClassNotFoundException =>
        e.printStackTrace()
        throw new IllegalStateException("deserialization send msg failed")
    }
  }

  private def deserializeMergeMsg[VD, ED, MSG](mergeMsgFilePath: String, conf: GraphXConf[VD, ED]): (MSG,MSG) => MSG = {
    try {
      val res = SerializationUtils.read(mergeMsgFilePath).asInstanceOf[(MSG, MSG) => MSG]
      res
    } catch {
      case e: ClassNotFoundException =>
        e.printStackTrace()
        throw new IllegalStateException("deserialization merge msg failed")
    }
  }
}
